sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/external/JdbcExternalWrite.java

```diff
@@ -0,0 +1,155 @@
+        JdbcExternalWrite.Configuration configuration) {
+      DataSourceConfiguration dataSourceConfiguration =
+          DataSourceConfiguration.create(configuration.driverClassName, configuration.jdbcUrl)
+              .withUsername(configuration.username)
+              .withPassword(configuration.password)
+              .withConnectionProperties(configuration.connectionProperties);
+
+      if (configuration.connectionInitSqls != null) {
+        List<String> connectionInitSqls =
+            StreamSupport.stream(configuration.connectionInitSqls.spliterator(), false)
+                .collect(Collectors.toList());
+        dataSourceConfiguration =
+            dataSourceConfiguration.withConnectionInitSqls(connectionInitSqls);
+      }
+
+      JdbcIO.PreparedStatementSetter<Row> preparedStatementSetter = new PrepareStatementXlangRow();
+
+      return JdbcIO.<Row>write()
+          .withDataSourceConfiguration(dataSourceConfiguration)
+          .withStatement(configuration.statement)
+          .withPreparedStatementSetter(preparedStatementSetter);
+    }
+
+    static class PrepareStatementXlangRow implements JdbcIO.PreparedStatementSetter<Row> {
+      @Override
+      public void setParameters(Row row, PreparedStatement statement) throws SQLException {
+        List<Schema.Field> fieldTypes = row.getSchema().getFields();
+        for (int i = 0; i < fieldTypes.size(); ++i) {
+          Schema.FieldType type = fieldTypes.get(i).getType();
+          if (Schema.FieldType.BYTE.equals(type)) {
+            statement.setByte(i + 1, row.getByte(i));
+          } else if (Schema.FieldType.INT16.equals(type)) {
+            statement.setShort(i + 1, row.getInt16(i));
+          } else if (Schema.FieldType.INT32.equals(type)) {
+            statement.setInt(i + 1, row.getInt32(i));
+          } else if (Schema.FieldType.INT64.equals(type)) {
+            statement.setLong(i + 1, row.getInt64(i));
+          } else if (Schema.FieldType.FLOAT.equals(type)) {
+            statement.setFloat(i + 1, row.getFloat(i));
+          } else if (Schema.FieldType.DOUBLE.equals(type)) {
+            statement.setDouble(i + 1, row.getDouble(i));
+          } else if (Schema.FieldType.DECIMAL.equals(type)) {
+            statement.setBigDecimal(i + 1, row.getDecimal(i));
+          } else if (Schema.FieldType.BOOLEAN.equals(type)) {
+            statement.setBoolean(i + 1, row.getBoolean(i));
+          } else if (Schema.FieldType.STRING.equals(type)) {
+            statement.setString(i + 1, row.getString(i));
+          } else if (Schema.FieldType.BYTES.equals(type)) {
+            statement.setBytes(i + 1, row.getBytes(i));
+          } else if (Schema.FieldType.DATETIME.equals(type)) {
+            statement.setDate(i + 1, new Date(row.getDateTime(i).toDateTime().getMillis()));
+          } else {
+            throw new IllegalArgumentException(String.format("Unknown FieldType: %s", type));
+          }
```

Comment:
nit: this could be a `switch` statement on `type`

Reply:
I tried it, but type isn't an enum or int - it's an instance of FieldType class

Reply:
Oh sorry I think you'd need to access the type name, `type.getTypeName()`, there's an example in https://github.com/apache/beam/blob/d851e6cae36d9acee845ca24a4cf50ff5ce704b4/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaTranslation.java#L84-L180

Reply:
I did not come out with an idea that there can be some enum under the hood. I've parsed atomic types only and left TODO to use writeRows with appropriate Jira issue https://issues.apache.org/jira/browse/BEAM-10396